backends/core/models.py

```diff
@@ -267,105 +355,222 @@ def get_ready(self):
+    return self._enqueued_task_count()
+
+  def _cancel_tasks(self):
+    key = self._get_prefixed_cache_key(CACHE_KEY_LIST_OF_TASKS_ENQUEUED)
+    enqueued_tasks = TaskEnqueued.where(task_namespace=key)
+    if enqueued_tasks:
+      tasks = [taskqueue.Task(name=t.task_name) for t in enqueued_tasks]
+      taskqueue.Queue().delete_tasks(tasks)
+      TaskEnqueued.where(task_namespace=key).delete()
+
+  def _enqueued_task_count(self):
+    key = self._get_prefixed_cache_key(CACHE_KEY_LIST_OF_TASKS_ENQUEUED)
+    return TaskEnqueued.count_in_namespace(key)
+
+  def _start_condition_is_fulfilled(self, start_condition):
+    preceding_job_status = start_condition.preceding_job.get_status()
+    if start_condition.condition == StartCondition.CONDITION.SUCCESS:
+      if preceding_job_status == Job.STATUS.FAILED:
+        return False
+    elif start_condition.condition == StartCondition.CONDITION.FAIL:
+      if preceding_job_status == Job.STATUS.SUCCEEDED:
+        return False
     return True
 
+  def start(self, max_retries=cache.MEMCACHE_DEFAULT_MAX_RETRIES):
+    """
+    Returns: Task object that was added to the task queue, otherwise None.
+    """
+    # Validates that preceding jobs fulfill the starting conditions.
+    for start_condition in self.start_conditions:
+      if self._start_condition_is_fulfilled(start_condition):
+        if start_condition.preceding_job.get_status() not in [
+            Job.STATUS.SUCCEEDED,
+            Job.STATUS.FAILED]:
+          return None
+      else:
+        # pipeline failure
+        self.set_status(Job.STATUS.FAILED)
+        self.pipeline.update(status=Pipeline.STATUS.FAILED,
+                             status_changed_at=datetime.now())
+        self.pipeline.stop()
+        return None
+
+    if self.pipeline.status == Pipeline.STATUS.FAILED:
+      return None
+
+    # Run the job with a concurrent-safe lock.
+    retries = 0
+    key = self._get_prefixed_cache_key(CACHE_KEY_STATUS)
+    while retries < max_retries:
+      curr_status = cache.get_memcache_client().get(key, for_cas=True)
+      if curr_status != Job.STATUS.WAITING:
+        return None
+      elif cache.get_memcache_client().cas(key, Job.STATUS.RUNNING,
+          time=cache.MEMCACHE_DEFAULT_EXPIRATION_TIME_SECONDS):
+        # Update the database status.
+        self.update(status=Job.STATUS.RUNNING, status_changed_at=datetime.now())
+        return self.run()
+      else:
+        retries += 1
+
+    # Failed to start.
+    from core import cloud_logging
+    cloud_logging.logger.log_struct({
+        'labels': {
+            'pipeline_id': self.pipeline_id,
+            'job_id': self.id,
+            'worker_class': self.worker_class,
+        },
+        'log_level': 'ERROR',
+        'message': 'Cannot start the job after multiple retries.',
+    })
+    return None
+
   def run(self):
+    # TODO remove the enqueued_workers_count field (since we use memcache for that purpuse)
```

Comment:
Why do we keep enqueued_workers_count at all?

Reply:
Removed.